testing/src/scenario/state.py

```diff
@@ -122,26 +122,40 @@ class _MaxPositionalArgs:
 
         _max_positional_args = n
 
-        def __new__(cls, *args: Any, **kwargs: Any):
+        @classmethod
+        def _annotate_class(cls):
+            """Record information about which parameters are positional vs. keyword-only."""
+            if hasattr(cls, "_init_parameters"):
+                # We don't support dynamically changing the signature of a
+                # class, so we assume here it's the same as previously.
+                return
             # inspect.signature guarantees the order of parameters is as
             # declared, which aligns with dataclasses. Simpler ways of
             # getting the arguments (like __annotations__) do not have that
             # guarantee, although in practice it is the case.
-            parameters = inspect.signature(cls.__init__).parameters
-            required_args = [
+            cls._init_parameters = parameters = inspect.signature(
+                cls.__init__
+            ).parameters
+            cls._init_kw_only = {
+                name
+                for name in tuple(parameters)[cls._max_positional_args :]
```

Comment:
I'm assuming this was auto-formatted by ruff, the dangling `:]` is kinda funny :]

Reply:
It was, yeah - I lazily just wrote it out on one line and let `tox -e fmt` clean up.

Does the trailing happy face bother you? I could do something like move `tuple(parameters)` out to a separate line to get the dict comprehension to be on a single line, if that would be better.